Std/Tactic/TryThis.lean

```diff
@@ -147,3 +148,52 @@ def addTermSuggestion (ref : Syntax) (e : Expr)
     (origSpan? : Option Syntax := none) : TermElabM Unit := do
   addSuggestion ref (← delabToRefinableSyntax e)
     (suggestionForMessage? := e) (origSpan? := origSpan?)
+
+open Lean Elab Elab.Tactic PrettyPrinter Meta Std.Tactic.TryThis
+
+/-- Add a suggestion for `have : t := e`. -/
+def addHaveSuggestion (ref : Syntax) (t? : Option Expr) (e : Expr)
+  (origSpan? : Option Syntax := none) : TermElabM Unit := do
+  let estx ← delabToRefinableSyntax e
+  let prop ← isProp (← inferType e)
+  let tac ← if let some t := t? then
+    let tstx ← delabToRefinableSyntax t
+    if prop then
+      `(tactic| have : $tstx := $estx)
+    else
+      `(tactic| let this : $tstx := $estx)
+  else
+    if prop then
+      `(tactic| have := $estx)
+    else
+      `(tactic| let this := $estx)
+  addSuggestion ref tac none origSpan?
+
+open Lean.Parser.Tactic
+open Lean.Syntax
+
+/-- Add a suggestion for `rw [h₁, ← h₂] at loc`. -/
+def addRewriteSuggestion (ref : Syntax) (rules : List (Expr × Bool))
+  (type? : Option Expr := none) (loc? : Option Expr := none)
+  (origSpan? : Option Syntax := none) :
+    TermElabM Unit := do
+  let rules_stx := TSepArray.ofElems <| ← rules.toArray.mapM fun ⟨e, symm⟩ => do
+    let t ← delabToRefinableSyntax e
+    if symm then `(rwRule| ← $t:term) else `(rwRule| $t:term)
+  let tac ← do
+    let loc ← loc?.mapM fun loc => do `(location| at $(← delab loc):term)
+    `(tactic| rw [$rules_stx,*] $(loc)?)
+
+  let mut tacMsg :=
+    let rulesMsg := MessageData.sbracket <| MessageData.joinSep
+      (rules.map fun ⟨e, symm⟩ => (if symm then "← " else "") ++ m!"{e}") ", "
+    if let some loc := loc? then
+      m!"rw {rulesMsg} at {loc}"
+    else
+      m!"rw {rulesMsg}"
```

Comment:
```suggestion
  let mut tacMsg := m!"{tac}"
```

Reply:
The reason for duplicating the structure is so that the message data contains embedded `Expr`s instead of just `Syntax`. (Yes, this is a hack and ideally there would be some better way to attach elaboration info to `Syntax` in a `MessageData`, but alas.)

Reply:
Thanks. I've revert this change and added a comment explaining why we're rebuilding the message.